src/client/static/islet.ts

```diff
@@ -0,0 +1,140 @@
+ */
+const whenIdle = (callback: () => void) => {
+  if ('requestIdleCallback' in window) {
+    window.requestIdleCallback(callback, { timeout: 500 });
+  } else {
+    setTimeout(callback, 300);
+  }
+};
+
+/**
+ * Use `whenVisible` to delay execution of the callback until a given element is visible
+ * in the viewport.
+ *
+ * Uses `IntersectionObserver` to determine when the element is visible.
+ * If `IntersectionObserver` is not supported use `setTimeout` failover to call back at the end of the call stack
+ *
+ * @param element : The html element that we want to observe;
+ * @param callback : This is fired when the element is visible in the viewport
+ */
+const whenVisible = (element: HTMLElement, callback: () => void) => {
+  if ('IntersectionObserver' in window) {
+    const io = new IntersectionObserver(([entry]) => {
+      if (!entry.isIntersecting) return;
+      // Disconnect this IntersectionObserver once seen
+      io.disconnect();
+      callback();
+    });
+
+    io.observe(element);
+  } else {
+    // IntersectionObserver is not supported so failover to calling back at the end of the call stack
+    setTimeout(() => callback(), 0);
+  }
+};
+
+/**
+ * This function dynamically imports, and then hydrates a specific component in
+ * a specific part of the page, or executes the named method from the imported
+ * script.
+ *
+ * @param name The name of the component we want to hydrate/named function and script we want to execute.
+ * @param type Determine if we want to hydrate a `component` or execute a named function from a `script`.
+ * @param element The location on the DOM where the component to hydrate exists
+ * @param props The props to pass to the component or arguments to the function
+ */
+const performImport = (
+  name: string,
+  type: string,
+  element: HTMLElement,
+  props?: Record<string, unknown>,
+) => {
+  let folder: string;
+
+  if (type === 'component') {
+    folder = 'components';
+  } else if (type === 'script') {
+    folder = 'lib';
+  } else {
+    throw new Error(`Unknown islet type: ${type}`);
+  }
+
+  import(
+    /* webpackInclude: /\.importable\.(m?)(j|t)s(x?)/ */
+    `../${folder}/${name}.importable`
+  )
+    .then((module) => {
+      if (type === 'component') {
+        console.log(`Hydrating island ${name}`);
+        hydrate(createElement(module[name], props), element);
+      } else if (type === 'script') {
+        console.log(`Script island ${name}`);
+        module[name](props);
+      }
+
+      element.setAttribute('data-gu-ready', 'true');
+    })
+    .catch((error) => {
+      if (name && error.message.includes(name)) {
+        console.error(
+          `🚨 Error importing ${name}. Components must live in the root of /components and follow the [MyComponent].importable.(m?)(j|t)s(x?) naming convention 🚨`,
+        );
```

Comment:
Do we want these errors  and logs to appear in non-development / staging environments?

Reply:
If you want logging in prod, I would recommend using `@guardian/libs`’s [logger](https://github.com/guardian/libs/tree/main/src/logger)